src/Snowflake/Command.php

```diff
@@ -149,18 +226,270 @@ public static function cloneDatabaseFromShared(Connection $connection, array $da
+                            'SHOW FUTURE GRANTS TO ROLE %s;',
+                            $role['name']
+                        )),
+                    ]
+                );
+            }
+        }
+        return $tmp;
+    }
+
+    private static function assignWorkspacesUsersAndRoles(Connection $connection, array $roles, array $users): array
+    {
+        foreach ($roles as $role) {
+            $grantsToRole = $connection->fetchAll(sprintf(
+                'SHOW GRANTS TO ROLE %s;',
+                $role['name']
+            ));
+
+            $ownershipToRole = array_filter($grantsToRole, fn($v) => $v['privilege'] === 'OWNERSHIP');
+
+            $rolesInRole = array_filter($ownershipToRole, fn($v) => $v['granted_on'] === 'ROLE');
+            $filteredRolesInRole = array_filter($rolesInRole, fn($v) => !in_array($v['name'], $roles));
+            $roles = array_merge($roles, array_combine(
+                array_map(fn($v) => $v['name'], $filteredRolesInRole),
+                $filteredRolesInRole
+            ));
+
+            $usersInRole = array_filter($ownershipToRole, fn($v) => $v['granted_on'] === 'USER');
+            $filteredUsersInRole = array_filter($usersInRole, fn($v) => !in_array($v['name'], $users));
+            $users = array_merge($users, array_map(fn($v) => $v['name'], $filteredUsersInRole));
+        }
+
+        return ['roles' => $roles, 'users' => $users];
+    }
+
+    public static function grantRoleToUser(Connection $connection, string $user, string $role): void
+    {
+        $connection->query(sprintf(
+            'GRANT ROLE %s TO USER %s',
+            QueryBuilder::quoteIdentifier($role),
+            QueryBuilder::quoteIdentifier($user)
+        ));
+    }
+
+    public static function assignGrantToRole(Connection $connection, array $grant): void
+    {
+        self::useRole($connection, $grant['granted_by']);
+
+        if ($grant['privilege'] === 'USAGE' && $grant['granted_on'] === 'ROLE') {
+            $sql = sprintf(
+                'GRANT %s %s TO %s %s %s',
+                $grant['granted_on'],
+                $grant['name'],
+                $grant['granted_to'],
+                $grant['grantee_name'],
+                $grant['grant_option'] === 'true' ? 'WITH GRANT OPTION' : '',
+            );
+        } else {
+            $sql = sprintf(
+                'GRANT %s ON %s %s TO %s %s %s',
+                $grant['privilege'],
+                $grant['granted_on'],
+                $grant['granted_on'] !== 'ACCOUNT' ? $grant['name'] : '',
+                $grant['granted_to'],
+                $grant['grantee_name'],
+                $grant['grant_option'] === 'true' ? 'WITH GRANT OPTION' : '',
+            );
+        }
+
+        $connection->query($sql);
+    }
+
+    public static function createMainRole(
+        Connection $connection,
+        string $warehouse,
+        array $users
+    ): void {
+        $mainRole = 'KEBOOLA_STORAGE';
```

Comment:
měl bys vytahnout z accountu - nemusí být tahle

Reply:
ach sooo... ok

Reply:
myslím, že v US je to `SAPI`

Reply:
nenapadá mě ale jak tu main roli vytáhnout... buď jí pošleme v configu (tak to mám i připravené) a nebo to prostě dáme do kódu podle stacku (což se mi moc nechce)

Reply:
Bude to owner té projektové role nebo uživatele.

Reply:
Na testingu:

```
OWNERSHIP	USER	CT_US_KEBOOLA_570165	ROLE	KEBOOLA_STORAGE
```

Reply:
to ano to máš pravdu, ale přes jakou query na to přijdeš, krom toho že budu zkoušet jednu roli po druhé

vem si že jsem přihlášený jako ACCOUNTADMIN a nevím na jakou roli se mám podívat abych to zjistil a ani nejde zjistit kdo vlastní mojí roli...

Reply:
šlo by to leda takhle
<img width="1104" alt="image" src="https://user-images.githubusercontent.com/12143866/214913626-581fde26-1b13-4470-9d3c-c1637ccc6aec.png">

Reply:
https://github.com/keboola/project-migration-tool/pull/1/commits/b894ce78f1ba1fb5f90d71898568e7e1b2d81edb

Reply:
Dopsal jsem tam komentář ještě...

Reply:
udělal jsem jak jsme se domluvili přes SHOW GRANT ON ROLE - https://github.com/keboola/project-migration-tool/pull/1/commits/35722d9ac923acbc7397f81b73347ddb4fdc8be1